torch/nested/_internal/nested_tensor.py

```diff
@@ -21,6 +21,7 @@ def get_tensor_id(tensor, *, coeff=1):
 class NestedTensor(torch.Tensor):
     _values: torch.Tensor  # type: ignore[assignment]
     _offsets: torch.Tensor
+    _lengths: Optional[torch.Tensor]
```

Comment:
What do we expect the .sizes() to return for this tensor?

Reply:
this should be the same size as `_offsets`

Reply:
Oh what I mean is what the .size() is for the NestedTensor
Currently, given a tensor with real size of [3, [1, 2, 3], 2]], values is represented as [1+2+3, 2] = [6, 2].
Today, in order to represent the ragged [1, 2, 3] dimension, we have a special int which we call "singleton int" returned in the size, e.g. [3, j0, 2] where the singleton int  j0 corresponds 1:1 with offsets Tensor objects. 
The semantics of this special int are basically that the main supported operation is `eq` and most other operations like `add` are disallowed. The idea is to allow this int to represent a kind of id for the offset and allow size comparison in O(1). 
Introducing lengths means that we might want a second singleton int corresponding 1:1 with lengths, and that we want to return that singleton int instead. 
There's some more context in `c10/core/SingletonSymNodeImpl.h`

Reply:
I see... I'll change it then

Reply:
@soulitzer good points! The increase in complexity scares me though.

Does it make sense to expand the current singleton symint to represent the combination of `(offsets, lengths)` when both are present?

Reply:
from offline discussion: when `lengths` exist, we should use those to produce the singleton symint instead of using `offsets`. Two jagged layout NTs with the same `lengths` will have the same singleton symint for that shape, and theoretically are compatible for binary op invocation. The kernel that performs the binary op must take into account both `offsets` and `lengths` to do the computation properly.